File: frontend/src/types/usage.ts
export interface CurrencyCost {
  currency: string;
  amount: number;
}

export interface UsageHeroOverview {
  totalTokens: number;
  totalInputTokens: number;
  totalOutputTokens: number;
  cacheReadTokens: number;
  cacheWriteTokens: number;
  reasoningTokens: number;
  requestsCount: number;
  costsByCurrency: CurrencyCost[];
  costCoveredTokens: number;
  costCoverageRatio: number;
  costCoveredRequests: number;
}

export interface UsageDailyTrendBucket {
  date: string;
  totalTokens: number;
  requestsCount: number;
  inputTokens: number;
  outputTokens: number;
  cacheTokens: number;
  costsByCurrency: CurrencyCost[];
}

export interface UsageModelBreakdown {
  model: string;
  provider: string;
  requestsCount: number;
  totalTokens: number;
  inputTokens: number;
  outputTokens: number;
  reasoningTokens: number;
  cacheTokens: number;
  cost: number | null;
  currency: string;
  costBasis: string;
}

export interface UsageSourceBreakdown {
  adapterId: string;
  sourceId: string;
  sourceName: string;
  appName: string;
  requestsCount: number;
  totalTokens: number;
  inputTokens: number;
  outputTokens: number;
  cacheTokens: number;
  costsByCurrency: CurrencyCost[];
  status: string;
  diagnostics: string[];
}

export interface UsageDateBreakdown {
  date: string;
  requestsCount: number;
  totalTokens: number;
  inputTokens: number;
  outputTokens: number;
  cacheTokens: number;
  costsByCurrency: CurrencyCost[];
}

export interface UsageSourceDiagnostic {
  sourceId: string;
  adapterId: string;
  level: string;
  message: string;
  timestamp: string;
}

export interface UsageScanStatus {
  scannedSourcesCount: number;
  totalEventsCount: number;
  lastScannedAt: string | null;
  activeScanTaskId: string | null;
  sourceDiagnostics: UsageSourceDiagnostic[];
}

export interface UsageDashboardDto {
  hero: UsageHeroOverview;
  dailyTrend: UsageDailyTrendBucket[];
  models: UsageModelBreakdown[];
  sources: UsageSourceBreakdown[];
  dates: UsageDateBreakdown[];
  scanStatus: UsageScanStatus;
}

export interface UsageDashboardFilter {
  timeRange?: string | null;
  adapterId?: string | null;
  sourceId?: string | null;
  model?: string | null;
  timezoneOffsetMinutes?: number | null;
}

export interface UsageScanOptions {
  mode?: "incremental" | "full" | null;
  sourceId?: string | null;
  adapterId?: string | null;
}

export function formatChineseTokens(tokens: number, precision = 2): string {
  if (!tokens || isNaN(tokens) || tokens === 0) {
    return "0";
  }
  const abs = Math.abs(tokens);
  const sign = tokens < 0 ? "-" : "";
  if (abs >= 100_000_000) {
    const val = (abs / 100_000_000).toFixed(precision);
    return `${sign}${val} 亿`;
  }
  if (abs >= 10_000) {
    const val = (abs / 10_000).toFixed(precision > 1 ? 1 : precision);
    return `${sign}${val} 万`;
  }
  return `${sign}${abs.toLocaleString()}`;
}

export function formatUsageCosts(costs: CurrencyCost[]): string {
  if (!costs || costs.length === 0) return "$0.00";
  return costs
    .map((c) => {
      const sym = c.currency.toUpperCase() === "CNY" ? "¥" : "$";
      return `${sym}${c.amount.toFixed(4).replace(/0+$/, "").replace(/\.$/, "")}`;
    })
    .join(" / ");
}

// Backward-compatibility aliases
export type ConversationUsageDashboard = UsageDashboardDto;
export type ConversationUsageScanStatus = UsageScanStatus;
export type ConversationUsageQuery = UsageDashboardFilter;
export type UsageTotals = UsageHeroOverview;
export type ModelUsageSummary = UsageModelBreakdown;
export type SourceUsageSummary = UsageSourceBreakdown;
export type DailyUsageTrendPoint = UsageDailyTrendBucket;

File: src-tauri/src/backend/dto/usage.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CurrencyCostDto {
    pub(crate) currency: String,
    pub(crate) amount: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageDailyTrendBucketDto {
    pub(crate) date: String,
    pub(crate) total_tokens: i64,
    pub(crate) requests_count: i64,
    pub(crate) input_tokens: i64,
    pub(crate) output_tokens: i64,
    pub(crate) cache_tokens: i64,
    pub(crate) costs_by_currency: Vec<CurrencyCostDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageModelBreakdownDto {
    pub(crate) model: String,
    pub(crate) provider: String,
    pub(crate) requests_count: i64,
    pub(crate) total_tokens: i64,
    pub(crate) input_tokens: i64,
    pub(crate) output_tokens: i64,
    pub(crate) reasoning_tokens: i64,
    pub(crate) cache_tokens: i64,
    pub(crate) cost: Option<f64>,
    pub(crate) currency: String,
    pub(crate) cost_basis: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageSourceBreakdownDto {
    pub(crate) adapter_id: String,
    pub(crate) source_id: String,
    pub(crate) source_name: String,
    pub(crate) app_name: String,
    pub(crate) requests_count: i64,
    pub(crate) total_tokens: i64,
    pub(crate) input_tokens: i64,
    pub(crate) output_tokens: i64,
    pub(crate) cache_tokens: i64,
    pub(crate) costs_by_currency: Vec<CurrencyCostDto>,
    pub(crate) status: String,
    pub(crate) diagnostics: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageDateBreakdownDto {
    pub(crate) date: String,
    pub(crate) requests_count: i64,
    pub(crate) total_tokens: i64,
    pub(crate) input_tokens: i64,
    pub(crate) output_tokens: i64,
    pub(crate) cache_tokens: i64,
    pub(crate) costs_by_currency: Vec<CurrencyCostDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageSourceDiagnosticDto {
    pub(crate) source_id: String,
    pub(crate) adapter_id: String,
    pub(crate) level: String,
    pub(crate) message: String,
    pub(crate) timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageHeroOverviewDto {
    pub(crate) total_tokens: i64,
    pub(crate) total_input_tokens: i64,
    pub(crate) total_output_tokens: i64,
    pub(crate) cache_read_tokens: i64,
    pub(crate) cache_write_tokens: i64,
    pub(crate) reasoning_tokens: i64,
    pub(crate) requests_count: i64,
    pub(crate) costs_by_currency: Vec<CurrencyCostDto>,
    pub(crate) cost_covered_tokens: i64,
    pub(crate) cost_coverage_ratio: f64,
    pub(crate) cost_covered_requests: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageScanStatusDto {
    pub(crate) scanned_sources_count: usize,
    pub(crate) total_events_count: usize,
    pub(crate) last_scanned_at: Option<String>,
    pub(crate) active_scan_task_id: Option<String>,
    pub(crate) source_diagnostics: Vec<UsageSourceDiagnosticDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageDashboardDto {
    pub(crate) hero: UsageHeroOverviewDto,
    pub(crate) daily_trend: Vec<UsageDailyTrendBucketDto>,
    pub(crate) models: Vec<UsageModelBreakdownDto>,
    pub(crate) sources: Vec<UsageSourceBreakdownDto>,
    pub(crate) dates: Vec<UsageDateBreakdownDto>,
    pub(crate) scan_status: UsageScanStatusDto,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageDashboardFilter {
    #[serde(default)]
    pub(crate) time_range: Option<String>,
    #[serde(default)]
    pub(crate) adapter_id: Option<String>,
    #[serde(default)]
    pub(crate) source_id: Option<String>,
    #[serde(default)]
    pub(crate) model: Option<String>,
    #[serde(default)]
    pub(crate) timezone_offset_minutes: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UsageScanOptions {
    #[serde(default)]
    pub(crate) mode: Option<String>, // "incremental" or "full"
    #[serde(default)]
    pub(crate) source_id: Option<String>,
    #[serde(default)]
    pub(crate) adapter_id: Option<String>,
}
